src/lib/InstantSearch.ts

```diff
@@ -234,33 +223,29 @@ See ${createDocumentationLink({
       this._searchFunction = searchFunction;
     }
 
-    const defaultRoutingOptions = {
-      stateMapping: simpleStateMapping(),
-      router: historyRouter(),
-    };
-
-    let routingOptions: Routing | null = null;
-    if (routing === true) {
-      routingOptions = defaultRoutingOptions;
-    } else if (isPlainObject(routing)) {
-      routingOptions = {
-        ...defaultRoutingOptions,
-        ...routing,
-      };
+    if (routing) {
+      const routerOptions = typeof routing === 'boolean' ? undefined : routing;
+      this.use(createRouter(routerOptions));
     }
+  }
 
-    if (routingOptions) {
-      this._routingManager = new RoutingManager({
-        ...routingOptions,
-        instantSearchInstance: this,
-      });
+  public use(...middleware: Middleware[]): this {
+    const newMiddlewareList = middleware.map(fn => {
+      const newMiddleware = fn({ instantSearchInstance: this });
+      this.middleware.push(newMiddleware);
 
-      this._createURL = this._routingManager.createURL;
-      this._initialUiState = {
-        ...initialUiState,
-        ...this._routingManager.read(),
-      };
+      return newMiddleware;
+    });
+
+    // If the instance has already started, we directly subscribe the
+    // middleware so they're notified of changes.
+    if (this.started) {
+      newMiddlewareList.forEach(m => {
+        m.subscribe();
```

Comment:
does the middleware not have an unsubscribe possibility? (could be returned from the subscribe call, not sure if useful)

Reply:
Ah no, I see it's separate, makes sense!